### DISCUSSION TITLE
Allow users to provide configuration files for tools they use in their projects

### DISCUSSION THREAD

**Discussion by julie777** [pyscaffold/pyscaffold#608] [Ideas]
This is actually part of a larger discussion that I want to have, but we should keep this part focused as it relates to #354 and #336.

I would like the user to be able to add an environment variable or a setting in their global pyscaffold config named PYSCAFFOLD_TEMPLATE_HOME or something similar. For the purposes of this discussion it would contain a subdirectory ```config``` with templates for configurations files. 

An example usage:
When generating a project we create a .readthedocs.yml file. 
User X has different options they want in their .readthedocs.yml file for all projects they create. So they just copy .readthedocs.yml from and existing project into $PYSCAFFOLD_TEMPLATE_HOME/config/readthedocs.yml and whenever `putup` creates a new project it will use the user template (assuming the env variable is set). Also, it can use the user template to update existing projects.

I'm not sure how we determine whether generated files or templates have changed and handle update and/or merging them, but that doesn't affect thee above proposal.

Note: that the template files are in a config directory and don't start with a ".", similar to the config files in the user's $HOME/.config directory

In addition, you will see in my example that not all config files use the config format. Tools use a variety of file names and file types for configuration. If the user chooses to have most configuration for tools put in pyproject.toml we can do so by copying from the templates, but we will need a tool "database" to define which tools can use pyproject.toml, and what file formats each tool supports.

I expect that when support for a new tools is added to pyscaffold that it will be done by creating an extension that
- defines configuration file information
- provides a default config template
- adds an entry to the tool database with information about support for pyproject.toml, etc.
- implements an adapter that allows pyscaffold to use it with a standard api for that tool type
- etc.

**Comment by abravalheri** [comment]
Hi @julie777, as you suggested, a PyScaffold extension is probably a good way of implementing this.

Something to keep in mind is that: all the extensions and templates we maintain in the github organisation are designed to work together.

For example, the CI templates "are aware" of which testenvs exist in the `tox.ini` file.

We also spend a lot of energy trying to avoid PyScaffold updates breaking too much stuff for users, and that depends on the contents of these templates.

---
Now I wonder if it is worthy to design an extension to work with generic templates, or just make it easier/clearer for users to use `pyscaffoldex-custom-extension` when they need it.

**Reply by julie777** [reply]
I understand what you are saying. 
Do the CI templates use the tox.ini, or are they separate?

In the past I have always tried to make it where the developer could run most of the CI locally, with only some potential combinations of platforms, etc. that were deferred to CI. This meant that CI broke less often.

(Again, I am trying to become part of the team and assimilate the collective knowledge about the project.)

**Reply by abravalheri** [reply]
> Do the CI templates use the tox.ini, or are they separate?

The files are separated but `tox` is the star! (and the CI depends on it).
The CI files work as wrappers for calling `tox`.

This is on purpose to do exactly what you suggested: keep the developer's flow in the local machine very similar to the one used in the CI.

For example: locally I run `tox` for my tests, `tox -e docs` for my docs, `tox -e build` for packaging things up and `tox -e publish` for publishing to PyPI.

The CI will run the same commands, only differing in some extra parameters that are more appropriate for the CI environment (for example for producing XML coverage files).

**Comment by julie777** [comment]
I am definitely proposing we put support for this in the core. I realize that it is an architectural change and that that would have to wait until the 5.0 release. I'll try to add more of my thoughts so you can see how you feel about it. Much of what I am suggesting really depends on the decisions about #592 and #601. 

Because I am new to the project, I am trying to understand what the rest of you have been thinking and planning and where the project is heading. Given that we aren't all sitting in a room talking, I am doing it with discussions, and trying to make each topic as specific as possible.

**Comment by abravalheri** [comment]
I am not confident about this approach. Although I come up myself with complex approaches from time to time 😅, I like to think that we should keep things easy and the implementation lean.

@FlorianWilhelm what are your thoughts about this?

Something that we can try @julie777 is to schedule a video call one weekend for example. It would be nice to be able to discuss directly and maybe very productive!

**Reply by julie777** [reply]
I think it would be nice to "meet" you both.   I am in Utah, USA so UTC -7, but I'm sure time zones won't be a problem.

**Reply by FlorianWilhelm** [reply]
Good idea, what about Sunday in a week, so 27th of February, 7 PM UTC+1 which should be 6 PM for @abravalheri and 12 PM for @julie777. Another week buys me some time to get up to speed and maybe reply more to some of the threads ;-)

**Reply by julie777** [reply]
On my calendar. What app do we use?

**Reply by FlorianWilhelm** [reply]
We could do Google Meet? Just take my mail from the git comits :-)

**Comment by FlorianWilhelm** [comment]
If I understand your suggestion correctly, @julie777, you are proposing something really powerful. Users could mix and match their own templates for configuration easily. I see the flexibility that comes with this approach for power users but would our target group really use a feature like this? I wonder even if one could implement a feature like this using vanilla cookie-cutter quite easily. Isn't one of the major selling points of PyScaffold that all our templates really play nicely together and come not only with sane defaults but with smart defaults? If we would allow users to mix and match their own templates directly, this could water down the actual purpose of PyScaffold, what it should be used for. 

For me, PyScaffold should be your imaginary senior developer friend, that comes to your desk and helps you set up a project by suggesting best practices (our templates) and even explaining them (our docs). Being a smart templating engine for general project setups is the niche of cookie-cutter.

That being said, I would still appreciate an extension but would be really hesitant to have it in core PyScaffold.

**Comment by julie777** [comment]
When I looked at cookie cutter it seemed way more complicated than necessary. The same with copier. They both try to be general enough and powerful enough to provide template instantiation for anything. We have a much more defined scope for our putup.

@FlorianWilhelm  you say "would our target group use a feature like this"

Another question to answer. What is our target group. I may have missed some of the answers to questions like this.
What is the goal of pyscaffold?
Who are our target users?

Part of my perspective about configuration comes from my experience in corporate jobs. Sometimes tools have been decided and you don't get a choice. I start a job and they use Mercurial and don't want to change. If we don't allow configuration then pyscaffold can't be used. Like it or not tools will be added by extension or otherwise. I have quite a few thoughts on this, but I think there is a better place for them.
I can't remember if this should be in another discussion or here. Some of them are overlapping.

**Comment by FlorianWilhelm** [comment]
The target group of PyScaffold used to be just me :-) and some colleagues at Blue Yonder. When working on some idea, some library, or whatnot, I no longer wanted to hack with Python scripts but instead, start with all the best practices there are out there in 5 seconds. So have all the benefits of CI/CD, versioning, autodoc, etc. right away and if my idea pans out distributing it is no problem at all. We decided to publish PyScaffold back then and other Python developers started to use it.

While developing PyScaffold and because of developing it, I learned a lot about the Python ecosystem and the best practices. I learned that it is smart to tie git tags to the actual version numbers of the distribution, to have an src folder from @RonnyPfannschmidt, and when @abravalheri joined many years ago, a lot about tox and the PEPs around pyproject.toml as well as many other things. So far @abravalheri and I have gone with the UNIX philosophy of integrating many powerful tools but without a wrapper around them (like what hatch and poetry do) and thus I guess our target group is people who want exactly that. People who rather say "I want to learn the best tools in the Python ecosystem so tell me which ones are the best" instead of people that say "I want to learn one new tool that does it all".  

Judging from the number of Github stars and downloads, PySaffold is not as popular as cookie-cutter, poetry, and hatch (stars only) but I feel that our user base is growing. We never really did much marketing, which would have been smart in hindsight, but who cares actually? PyScaffold covers a niche that the other tools don't cover and that's a good thing, right? Of course, it's cool to see that PyScaffold has many users but a huge user base never has been the goal, and shouldn't. Anderson and I more or less created it the way we want to use it. And of course, it's fun to develop it.

There is one more thing regarding the goals of PyScaffold. They should not be over-ambitious and thus time-consuming as Anderson and I work in our spare time on it. For this reason, I am always a fan of keeping it simple and stupid as this helps us with maintainability. 
You mentioned Mercurial. I think Anderson and I discussed it quite some time ago if it should be integrated, or maybe I discussed it with someone from Blue Yonder, nevertheless, having another version control system besides Git, which is used by like 95% of all developers, and keeping the feature parity would mean much, much more maintenance for us. So in this case, we are sorry for Mercurial users but one has to understand that pure maintenance has to be kept small in order to keep an OSS project by spare-time developers fun and alive. BTW, in the last years most of the work was done by Anderson and I spent only about 2-3 hours a week on average on it.  
These are just my 5 cents and please feel Anderson to contradict and comment.

**Comment by abravalheri** [comment]
Florian comment is, as always, very wise, and I think it captures perfectly the philosophy of the project.

Something that I would like to add is that PyScaffold, accidentally or not, have this *"educational"* aspect to it.

New Python developers can start coding and making use of all the nice tools of the ecosystem right away.
As they get more experienced, they can open the files we provide (e.g. `tox.ini`, `.pre-commit-hook.yml`, `.cirrus.yml`), start learning from the examples and personalise them.

While we try to hide complexity and facilitate things, I wouldn't go to far as wrapping all the tools and making it impossible for the user to learn how to use them on their own or to modify how they behave.

**Comment by julie777** [comment]
You both raise good points as always. The idea that pyscaffold has an educational aspect is part of what I want.
The fact that new developers can start developing right away and not have to learn a dozen tools first is what I want.

I think we are somewhat in agreement on some of the concepts, but not on the details.

> While we try to hide complexity and facilitate things, I wouldn't go to far as wrapping all the tools and making it impossible for the user to learn how to use them on their own or to modify how they behave

The above might be a good example. 
1. I don't want to completely wrap all tools and make it hard for the user to transition to using tools directly - when they are ready for that step - one tool at a time.
2. I want to educate them on the development lifecycle by providing and easy way to follow the steps to create and publish or deploy a project using best practices.
3. I don't want to force them to do all the steps right away. (Refer to my example of the new python developer creating a program for personal use, or to learn python)

I don't know if I have expressed it this way, but I see putup as one of the tools that implements one of the best practices. Git is another tool that implements best practices.

Another tool that can run all the required tools in simplified form to show the user how they should use the process and hold their hand as they get started is something that I want to put on the table. Pyscaffold is the master project, but there is no reason for us not to develop other tools, or allow others to do so and contribute.

(I've done a lot of educating and mentoring in my career, and it definitely colors my perspective about how to help developers learn and grow, becoming good developers.)

**Comment by julie777** [comment]
> The target group of PyScaffold used to be just me :-) and some colleagues at Blue Yonder.

Well this has certainly changed and I thank both of you @FlorianWilhelm and @abravalheri  for all of the work you have done.

Should we create discussions for: target group, goals of the project, and any other high level topics that we don't have written down with deisions?

I think the more focused we can keep each discussion thread the easier it will be to get to a conclusion and close the discussion.

**Comment by julie777** [comment]
> in the last years most of the work was done by Anderson and I spent only about 2-3 hours a week on average on it.

I suspected that you both were working regular jobs and that this was on personal time. 

I am not working a regular job with fixed hours, so I have much more time available to work on pysk in a typical week. That is why I keep starting discussions. I am trying to understand what you already know. I am trying see where my ideas complement yours. I am trying to see where I can best focus my efforts to make pysk better.
